drivers/eeprom/eeprom_mchp_xec.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2022 Microchip Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_xec_eeprom
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/eeprom.h>
+#include <soc.h>
+
+#ifdef CONFIG_PINCTRL
+#include <zephyr/drivers/pinctrl.h>
+#endif
+
+#define LOG_LEVEL CONFIG_EEPROM_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(eeprom_xec);
+
+/* EEPROM Mode Register */
+#define XEC_EEPROM_MODE_ACTIVATE		BIT(0)
+
+/* EEPROM Status Register */
+#define XEC_EEPROM_STS_TRANSFER_COMPL		BIT(0)
+
+/* EEPROM Execute Register - Transfer size bit position */
+#define XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS	(24)
+
+/* EEPROM Execute Register - Commands */
+#define XEC_EEPROM_EXC_CMD_READ			0x00000U
+#define XEC_EEPROM_EXC_CMD_WRITE		0x10000U
+#define XEC_EEPROM_EXC_CMD_READ_STS		0x20000U
+#define XEC_EEPROM_EXC_CMD_WRITE_STS		0x30000U
+
+/* EEPROM Execute Register - Address mask */
+#define XEC_EEPROM_EXC_ADDR_MASK		0x7FFU
+
+/* EEPROM Status Byte */
+#define XEC_EEPROM_STS_BYTE_WIP			BIT(0)
+#define XEC_EEPROM_STS_BYTE_WENB		BIT(1)
+
+/* EEPROM Read/Write Transfer Size */
+#define XEC_EEPROM_PAGE_SIZE			32U
+#define XEC_EEPROM_TRANSFER_SIZE_READ		XEC_EEPROM_PAGE_SIZE
+#define XEC_EEPROM_TRANSFER_SIZE_WRITE		XEC_EEPROM_PAGE_SIZE
+
+#define XEC_EEPROM_DELAY_US			500U
+#define XEC_EEPROM_DELAY_BUSY_POLL_US		50U
+#define XEC_EEPROM_XFER_COMPL_RETRY_COUNT	10U
+
+struct eeprom_xec_regs {
+	uint32_t mode;
+	uint32_t execute;
+	uint32_t status;
+	uint32_t intr_enable;
+	uint32_t password;
+	uint32_t unlock;
+	uint32_t lock;
+	uint32_t _reserved;
+	uint8_t buffer[XEC_EEPROM_PAGE_SIZE];
+};
+
+struct eeprom_xec_config {
+	struct eeprom_xec_regs * const regs;
+	size_t size;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pcfg;
+#endif
+};
+
+K_MUTEX_DEFINE(lock);
```

Comment:
Why isn't this part of a `struct eeprom_xec_data`?

Reply:
there is no `eeprom_xec_data`

Reply:
Please add one and move the mutex there. That is what the driver instance data is for. A struct wrapping a single member will not result in any overhead.